src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1208,10 +1200,38 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
           // trait T { def f: Object }; object O extends T { object f }. Need to generate method f in O.
           if (module.isOverridingSymbol) matchingInnerObject() else Nil
         else
-          newInnerObject()
+          newInnerObject(site, module)
       )
       transformTrees(newTrees map localTyper.typedPos(moduleDef.pos))
     }
+    def newInnerObject(site: Symbol, module: Symbol): List[Tree] = {
+      // Create the module var unless it is already in the module owner's scope.
+      // The lookup is on module.enclClass and not module.owner lest there be a
+      // nullary method between us and the class; see SI-5012.
+      val moduleVar = site newModuleVarSymbol module
+      val moduleNew = gen.newModule(module, moduleVar.tpe)
+      val body      = if (site.isTrait) moduleNew else {
+        if (moduleVar.enclClass == NoSymbol) gen.mkAssignAndReturn(moduleVar, moduleNew) // e.g. neg/t6666c.scala, will be an error later on.
+        else gen.mkAssignAndReturn(moduleVar, moduleNew) // will be wrapped in double checked lock idiom in lazyvals
```

Comment:
Ha!

In an earlier version of this change, I synthesized all the lazy val ASTs right here, which hit a `NoSymbol` related error in some code that would have otherwise been flagged as erroneous in a later phase. So I only did that in the second branch of this `if`.

But now, since I defer the work until `lazyvals`, the second branch of the `if` became identical to the first :)